source/grid/grid_in.cc

```diff
@@ -340,139 +340,172 @@ GridIn<dim, spacedim>::read_vtk(std::istream &in)
+                  else if (set == "ManifoldID")
+                    boundary_quad.manifold_id = static_cast<types::manifold_id>(id);
+                  else
+                    Assert(false, ExcInternalError());             
+                }
+                for (auto &boundary_line : subcelldata.boundary_lines)
+                {
+                  double id;
+                  in >> id;
+                  if (set == "MaterialID")
+                    boundary_line.material_id = static_cast<types::material_id>(id);
+                  else if (set == "ManifoldID")
+                    boundary_line.manifold_id = static_cast<types::manifold_id>(id);
+                  else
+                    Assert(false, ExcInternalError());
+                }
+              }
+              else if (dim == 2)
+              {
+                for (auto &boundary_line : subcelldata.boundary_lines)
+                {
+                  double id;
+                  in >> id;
+                  if (set == "MaterialID")
+                    boundary_line.material_id = static_cast<types::material_id>(id);
+                  else if (set == "ManifoldID")
+                    boundary_line.manifold_id = static_cast<types::manifold_id>(id);
+                  else
+                    Assert(false, ExcInternalError());             
+                }
               }
+              in >> keyword;
           }
+        }
+      
+      }
+      //////////////////
+      //// Addition of FIELD DATA:
+      //////////////////      
+      if(keyword == "FIELD")
+      {
+        unsigned int n_fields;
+        in >> keyword;
+        AssertThrow(
+          keyword == "FieldData",
+          ExcMessage("While reading VTK file, missing keyword FieldData"));
+        
+        in >> n_fields;
+        
+        for(unsigned int i = 0;i < n_fields;++i)
+        {
+          std::string section_name;
+          std::string data_type;
+          unsigned int temp,n_ids;
+          double data;
+          in>>section_name;
+          in>>temp;
+          in>>n_ids;
+          AssertThrow(n_ids == n_geometric_objects,
+                        ExcMessage("The VTK reader found a FIELD statement "
+                        "that lists a total of " + Utilities::int_to_string(n_ids) +
+                        " cell data objects, but this needs to equal the number of cells (which is " +
+                        Utilities::int_to_string(cells.size()) + ") plus the number of quads (" +
+                        Utilities::int_to_string(subcelldata.boundary_quads.size()) +
+                        " in 3d or the number of lines (" + Utilities::int_to_string(
+                          subcelldata.boundary_lines.size()) + ") in 2d."));
+          in>>data_type;
+          std::map<int,double> temp_data;
+          for (unsigned int j=0;j<n_ids;j++)
+          {
+            in>>data;
+            if (j<cells.size())
+              temp_data[j]=data;
```

Comment:
This is interesting, because you're looping over `j` and then you create a map from `j` to `data`. But `j` consists of indices 0, 1, 2, ... so that map is really quite simple: It consists of data `(0,a0), (1,a1), (2,a2), ...`. This can be stored in an easier way: You don't need the expensive map, you could just store it as a vector `(a0,a1,a2,...)`.